atspi-common/src/object_match.rs

```diff
@@ -0,0 +1,281 @@
+use std::{collections::HashMap, marker::PhantomData};
+
+use serde::{Deserialize, Serialize};
+use zvariant::{Signature, Type};
+
+use crate::{Interface, InterfaceSet, Role, StateSet};
+
+/// Defines how an object-tree is to be traversed.
+/// Used in `CollectionProxy`.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize, Type)]
+#[repr(u32)]
+pub enum TreeTraversalType {
+	/// When traversing the tree, restrict to children of the current object.
+	RestrictChildren,
+
+	/// When traversing the tree, restrict to siblings of the current object.
+	RestrictSibling,
+
+	/// Traverse the tree in order of appearance.
+	#[default]
+	Inorder,
+}
+
+/// Definition of match rules for accessible objects.
+/// Rule(s) against which we can match an  object or a collection thereof.
+///
+/// # Examples
+///  ```Rust
+///     let builder = MatchRule::builder();
+/// ```
+///
+#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ObjectMatchRule {
+	pub states: StateSet,
+	pub states_mt: MatchType,
+	pub attr: HashMap<String, String>,
+	pub attr_mt: MatchType,
+	pub roles: Vec<Role>,
+	pub roles_mt: MatchType,
+	pub ifaces: InterfaceSet,
+	pub ifaces_mt: MatchType,
+	pub invert: bool,
+	// Private phantom, gets compiled away.
+	// Here to ensure the builder is the only route to obtain a `MatchRule`
+	#[serde(skip)]
+	phantom: std::marker::PhantomData<()>,
+}
+
+// !!! WARNING !!! :
+//
+// State and Role are defined as u32 in Accessible.xml but as i32 in Collection.xml
+//
+// The signature on StateSet is defined inconsistently in the XMLs
+// Accessible.xml: GetState type="au"
+// Collection.xml: GetMatches argument name="rule" type="(aiia{ss}iaiiasib)"
+// The latter starts with ai, which is a signature for an array of signed ints i32.
+//
+// https://gitlab.gnome.org/federico/at-spi2-core/-/commit/4885efedeef71e0df8210622771a0b1bb10e194d
+impl Type for ObjectMatchRule {
+	fn signature() -> Signature<'static> {
+		Signature::from_str_unchecked("(aiia{ss}iaiiasib)")
+	}
+}
+
```

Comment:
Its implications are mostly hypothetical and admittedly remote. 

* `StateSet` 

StateSet is a bitfield. As long as each side sticks to bitwise operations only, I think we are all good.

In order for this to go off the rails, imagine the implementer of `Collection` does arithmetic operations on the constituents to retrieve individual states, then the implications depend on what exactly they it is they do, but from our perspective undefined behaviour.

* `Role`
I suspect the number of roles will never exceed 2^31, so no problems.

* Sentinel values
Not problematic in this particular case, because we are  sending a u32, but consider the habit of using -1 to encode something special.and receiving an i32, which you encode as u32.  In two's complement -1 = 11111111 11111111 11111111 11111111
Equals `u32::MAX`, which is many states (in a state set) or likely an illegal or unwanted value for an enum.

Reply:
My OCD makes me think protocols should be consistent across the board.

Reply:
> My OCD makes me think protocols should be consistent across the board.

Option types on DBus would help a lot....

Reply:
Could you explain how that would help here.